url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/HW4Main.java

```diff
@@ -0,0 +1,318 @@
+		String variables = "";
+		String values = "";
+		Solution found = new Solution(false, null);
+		
+		//build variables string
+		for (int i = 0; i < cp.size; i++)
+		{
+			if (variables=="") variables = variables + names.charAt(i);
+			else variables = variables + "." + names.charAt(i);
+		}
+		
+		//all possible values 
+		for (int i = 0; i < cp.width; i++)
+		{
+			for (int j=0; j<cp.height; j++)
+			{
+				if (values=="") values = "" + Rect.getInt(i, j, cp.width);
+				else values = values + "." + Rect.getInt(i, j, cp.width);
+			}
+		}
+		
+		DomainList dl = new DomainList(cp.size, 0);
+		
+		/*
+		 * The domain for each variable is:  {(x,y) | x in 0:m-w, y in 0:n-h } That is all x,y coordinates (for the upper left starting corner)
+    	 * that are for x between 0 and m-w and for y between 0 and n-h.  This is on a grid that has 0,0 in upper left.
+    	 * we proceed to fill the domain lists
+		 */
+		int ri = 0;
+		for (Rect r : cp.compList) //for each component
+		{
+			for (int x=0; x<cp.width; x++)
+			{
+				for (int y=0; y<cp.height; y++)
+				{
+					if (x <= cp.width - r.width && y <= cp.height - r.height)
+					{
+						dl.addValue(ri, Rect.getInt(x,y,cp.width));
+					}
+				}
+			}
+			ri++;
+		}
+		
+		
+		//printDL(dl,cp);
+
+		CSP csp = new CSP(variables, values);
+
+		// build the constraints
+		// ehaavat wbxr ng 41 jrfg jurrybpx: ebohfgarff vf sbe chffvrf
+		// TODO: make this simpler if not considering non-rectangular shapes
+		// TODO: make this work for problem 3
+		int ai = 0;
+		HashSet<Integer> seen = new HashSet<Integer>();
+		for (Rect a : cp.compList) //this is the first component of the binary constraint
+		{
+			Set<Integer> ad = dl.getValues(ai); //get the legal values domain for a
+			int bi=0;
+			for (Rect b : cp.compList) //we run the first component paired against every other component
+			{
+				String constraint = "";//we'll put the constraint string in here
+				if (ai != bi) // && !seen.contains(bi)) //ignore if the same index indicating the same component or if we've already looked at this pair in reverse
+				{			  // actually we do want symmetric constraints.
+					for (Integer i : ad)
+					{
+						Set<Integer> bd = dl.getValues(bi); //get a copy of the legal values domain for b
+						Set<Integer> aCols = coordSet(i, cp.width, a);
+						bd.removeAll(aCols); //remove all direct collision coordinates from domain for b
+						for (Integer c : bd) //now for all the possible positions for b that don't START in a
+						{					 //we need to check for start positions that would overlap in the other direction
+							Set<Integer> bCols = coordSet(c, cp.width,b);
+							if (java.util.Collections.disjoint((Collection<Integer>)bCols, (Collection<Integer>)aCols)) 
+							{//are any of the b coords in a's spots? if not great add the constraint!
+								if (constraint =="") constraint = i+"."+c;
+								else constraint = constraint + ":" + i+"."+c;
+							}
+						}
+					}
+				}
+				if (constraint!="")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.